### PR TITLE
fix(bamlfuzz): tolerate extra null keys from upstream BAML union serialization bug

### PR DESCRIPTION
**Pull request by invakid404**
## Problem

BAML's Go codegen ([boundaryml/baml#3690](https://github.com/boundaryml/baml/issues/3690)) serializes null-valued optional fields from a class union arm even when the active union arm is a map or a different class.

Example:
- Schema: `FuzzClass1 | map[string]int`, where `FuzzClass1` has `Fuzz_field_0: optional<float>`
- LLM returns: `{"k0": -26}`
- Expected (oracle): `{"k0": -26}`
- BAML produces: `{"Fuzz_field_0": null, "k0": -26}`

The bamlfuzz dynamic oracle flags the extra `Fuzz_field_0: null` as either a "schema key order mismatch" or a "semantic diff", accounting for 9/42 nightly #8 failures.

## Fix

Relax the **comparison-only** paths so a missing key is treated as equivalent to a null-valued key:

- **`envelope.go`** — `deepEqualSemantic` / `diffAny`: drop keys that are null-valued on one side and absent from the other before comparing lengths / emitting diff entries (helper `withoutExtraNullKeys`).
- **`order.go`** — `walkClass` and the `KindMap` case: strip leaked null keys (helper `stripExtraNullKeys`) before the wire-order equality check.

Both helpers are documented as a workaround for the upstream bug and marked for removal once it lands.

### What this does NOT touch

- No change to serialization or generated code (no `omitempty`).
- Non-null mismatches (extra/missing non-null keys, null-vs-value) still produce diffs.
- Only the presence of a leaked null key is suppressed.

## Tests

- `envelope_test.go`: `SemanticDiff` returns zero entries when the only difference is an extra null key; real differences still surface. Updated the `null vs missing` case in `TestSemanticEqual` to reflect the new contract.
- `order_test.go`: `SchemaOrderDiffWithChoices` returns zero diffs for leaked null keys at class and map nodes (including a test mirroring the exact `class | map<string,int>` shape from #3690); extra non-null keys still diff.

`GOWORK=off go test ./codegen/bamlfuzz/... -count=1` passes.

Umbrella: #349

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Ignore extra null-valued keys present only on the actual side for value comparisons and key-order checks, preventing spurious diffs while preserving genuine mismatches.

* **New Features**
  * Add a symmetric "parity" mode that tolerates extra nulls on either side for parity comparisons and order checks.

* **Tests**
  * Add extensive unit and integration tests validating asymmetric and parity null-key tolerance across diffs, semantic equality, schema-order, nested structures, and unions.

* **Integration**
  * Switch relevant integration checks to parity comparisons and make union-arm selection tolerant of leaked nulls.

### CODE REVIEW

File: adapters/common/codegen/bamlfuzz/order.go [hunk 1]
```diff
@@ -196,6 +241,83 @@ func (w *orderWalker) setErr(err error) {
...
+	case KindUnion:
+		for _, v := range typ.Variants {
+			if canBeNull(v) {
+				return true
+			}
+		}
+		return false
+	default:
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/bamlfuzz/order.go]: This is already handled. At `order.go:387`, `stripLeakedNulls := !CanBeNull(*typ.Inner)` — when the inner type IS nullable, `stripLeakedNulls` is `false`, so `orderKeys` receives `stripAllNulls=false` and falls through to the reference-aware `stripExtraNullKeys` path (not `stripNullKeys`). `stripExtraNullKeys` only strips keys that are null-valued AND absent from the reference side — for `map<string,null>` where both sides carry the same keys, nothing is stripped and the order comparison fires normally.

Covered by existing tests:
- `TestSchemaOrderDiffParity_MapNullValueTypeKeepsOrderCheck` (order_test.go:947) — `map<string,null>` with reordered keys expects one order diff
- `TestSchemaOrderDiffParity_MapOptionalInnerKeepsOrderCheck` (order_test.go:984) — same for `map<string, optional<int>>`

Finding 2 (from the duplicate review) is the same claim at the call site — also handled by the `!CanBeNull` gate.

@coderabbitai Both findings verified as already addressed in the current code.
